### ISSUE TITLE
RegisterConditional misbehaves in case of registration of IEnumerable<T> (container uncontrolled collections)

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#468]
Reported by @ensisnoctis:

I still have an inquiry towards the `IEnumerable<T>` part. Consider this code:

``` c#
public class Service<T>
{
    protected IEnumerable<T> Data;
    public Service(IEnumerable<T> data) => this.Data = data;
    public virtual IEnumerable<T> GetData => this.Data;
}

public class NumberService : Service<float>
{
    public NumberService(IEnumerable<float> data) : base(data) { } // A
}

public class StringService : Service<string>
{
    public StringService(IEnumerable<string> data) : base(data) { } // A
}

void Main()
{
    var ioc = new Container();
    ioc.RegisterCollection(typeof(Service<>), new[] { this.GetType().Assembly });
    ioc.RegisterConditional(typeof(IEnumerable<float>),
        Lifestyle.Singleton.CreateRegistration(() => new[] { 1.41f, 2.71f, 3.14f }, ioc), 
        context => true);   // A
    ioc.RegisterConditional(typeof(IEnumerable<string>), 
        Lifestyle.Singleton.CreateRegistration(() => new[] { "sqrt2", "e", "pi"  }, ioc), 
        context => false);  // A, C

    ioc.GetInstance(typeof(Service<float>)).Dump();
    ioc.GetInstance(typeof(Service<string>)).Dump();
}
```

You can see a base `Service` and its two implementations, both of which take literal constructor parameters. Those are supplied in `Main()`, conditionally.

Despite conditional prohibiting it (comment `C`), the second service gets resolved successfully. I did read this topic and understand that it's by design, I even got it to work by replacing all `IEnumerable<T>` with `T[]` array type (everywhere commented `A`).

My question though is this: because `IEnumerable<T>` is violating some principles here and there, why not just throw an exception explaining it instead of doing such an, in my humble opinion, unintuitive thing as to override the predicate (`C`) silently?

**Comment by dotnetjunkie** [comment]
I checked this, and this is a bug considering the registration of `IEnumerabe<T>`. My simplest advice is: don't register `IEnumerable<string>` and `IEnumerable<float>` since they are not components, but configuration values, but the report might be a simplification of what's going on.

Another way that Simple Injector breaks is by doing the following:

``` c#
ioc.RegisterConditional(
    typeof(IEnumerable<string>),
    Lifestyle.Singleton.CreateRegistration(() => new[] { "sqrt2", "e", "pi" }, ioc),
    context => false);

ioc.RegisterConditional(
    typeof(IEnumerable<string>),
    Lifestyle.Singleton.CreateRegistration(() => new[] { "foo", "bar", "foobar" }, ioc),
    context => false);
```
Here, Simple Injector throws an exception on the second registration (for the same `IEnumerable<string>`):

>System.InvalidOperationException: Collection of items for type String has already been registered and the container is currently not configured to allow overriding registrations. To allow overriding the current registration, please create the container using the constructor overload that takes a ContainerOptions instance and set the AllowOverridingRegistrations property to true. In case it is your goal to append items to an already registered collection, please use the AppendToCollection extension method. This method is located in the SimpleInjector.Advanced namespace.

**Comment by dotnetjunkie** [comment]
Fixed in [v4.0.12](https://www.nuget.org/packages/SimpleInjector/4.0.12).
